components/brave_federated/adapters/flower_helper.cc

```diff
@@ -0,0 +1,191 @@
+#include "brave/third_party/flower/src/proto/flwr/proto/fleet.pb.h"
+#include "brave/third_party/flower/src/proto/flwr/proto/node.pb.h"
+#include "brave/third_party/flower/src/proto/flwr/proto/transport.pb.h"
+
+namespace brave_federated {
+
+std::string BuildGetTasksPayload() {
+  std::string request;
+
+  flower::Node node;
+  node.set_node_id(0);
+  node.set_anonymous(true);
+
+  flower::PullTaskInsRequest pull_task_instructions_request;
+  *pull_task_instructions_request.mutable_node() = node;
+  pull_task_instructions_request.add_task_ids("0");
+  pull_task_instructions_request.SerializeToString(&request);
+
+  return request;
+}
+
+absl::optional<Task> ParseTask(const flower::TaskIns& task_instruction) {
+  const TaskId& task_id = {
+      .id = task_instruction.task_id(),
+      .group_id = task_instruction.group_id(),
+      .family_id = task_instruction.workload_id(),
+  };
+  if (!task_id.IsValid()) {
+    VLOG(2) << "Invalid task id received from FL service";
+    return absl::nullopt;
+  }
+
+  if (!task_instruction.has_task()) {
+    VLOG(2) << "Task object is missing from task instruction";
+    return absl::nullopt;
+  }
+  flower::Task flower_task = task_instruction.task();
+
+  if (!flower_task.has_legacy_server_message()) {
+    VLOG(2) << "Server message is missing from task object";
+    return absl::nullopt;
+  }
+  flower::ServerMessage message = flower_task.legacy_server_message();
+
+  Configs config;
+  TaskType type = TaskType::kUndefined;
+  std::vector<Weights> parameters;
+  if (message.has_fit_ins()) {
+    type = TaskType::kTraining;
+
+    if (!message.fit_ins().has_parameters()) {
+      VLOG(2) << "Parameters are missing from fit instruction";
+      return absl::nullopt;
+    }
+    parameters = GetVectorsFromParameters(message.fit_ins().parameters());
+    if (parameters.empty()) {
+      VLOG(2) << "Parameters vectors received from FL service are empty";
+      return absl::nullopt;
+    }
+    config = ConfigsFromProto(message.fit_ins().config());
+  } else if (message.has_evaluate_ins()) {
+    type = TaskType::kEvaluation;
+
+    if (!message.evaluate_ins().has_parameters()) {
+      VLOG(2) << "Parameters are missing from eval instruction";
+      return absl::nullopt;
+    }
+    parameters = GetVectorsFromParameters(message.evaluate_ins().parameters());
+    if (parameters.empty()) {
+      VLOG(3) << "Parameters vectors received from FL service are empty";
+      return absl::nullopt;
+    }
+    config = ConfigsFromProto(message.evaluate_ins().config());
+  } else {
+    VLOG(2) << "**: Received unrecognized instruction from FL service";
+    return absl::nullopt;
+  }
+
+  return Task(task_id, type, "token", parameters, config);
+}
+
+absl::optional<TaskList> ParseTaskListFromResponseBody(
+    const std::string& response_body) {
+  flower::PullTaskInsResponse response;
+  if (!response.ParseFromString(response_body)) {
+    VLOG(1) << "Failed to parse response body";
```

Comment:
Should we specify FL in the error message, to specify that FL parsing failed?
We used VLOG(2) for errors above, should we use the same log level here?

Reply:
Fixed here and in other places in the component 👍